server/src/main/java/org/elasticsearch/action/bulk/TransportAbstractBulkAction.java

```diff
@@ -180,12 +181,34 @@ protected void doRun() throws IOException {
     private boolean applyPipelines(Task task, BulkRequest bulkRequest, Executor executor, ActionListener<BulkResponse> listener)
         throws IOException {
         boolean hasIndexRequestsWithPipelines = false;
-        final Metadata metadata = clusterService.state().getMetadata();
-        Map<String, ComponentTemplate> templateSubstitutions = bulkRequest.getComponentTemplateSubstitutions();
+        final Metadata metadata;
+        Map<String, ComponentTemplate> componentTemplateSubstitutions = bulkRequest.getComponentTemplateSubstitutions();
+        if (bulkRequest.isSimulated() && componentTemplateSubstitutions.isEmpty() == false) {
+            /*
+             * If this is a simulated request, and there are template substitutions, then we want to create and use a new metadata that has
+             * those templates.
+             */
+            Metadata.Builder simulatedMetadataBuilder = Metadata.builder(clusterService.state().getMetadata());
+            if (componentTemplateSubstitutions.isEmpty() == false) {
+                Map<String, ComponentTemplate> updatedComponentTemplates = new HashMap<>();
+                updatedComponentTemplates.putAll(clusterService.state().metadata().componentTemplates());
+                updatedComponentTemplates.putAll(componentTemplateSubstitutions);
+                simulatedMetadataBuilder.componentTemplates(updatedComponentTemplates);
+            }
+            // We now remove the index from the simulated metadata to force the templates to be used
+            for (DocWriteRequest<?> actionRequest : bulkRequest.requests) {
+                IndexRequest indexRequest = getIndexWriteRequest(actionRequest);
+                simulatedMetadataBuilder.remove(indexRequest.index());
+            }
```

Comment:
> First of all, I want to verify an assumption, getIndexWriteRequest(actionRequest) can return null in the case of a delete request. Is it true that it's impossible to set a delete request via the simulate API? If yes, then this is ok, otherwise we might be facing an NPE.

It's only possible to simulate index requests. I can do a null check just in case that changes at some point in the future.

> Now to my second concern, what if the target is a data stream? This index removal from the simulated cluster state probably will have no effect, right?

Good catch. This is a real problem. I'll add some tests for it and fix it.

Reply:
Just a note: I added a test for data stream support to the yaml rest test rather than TransportSimulateBulkActionIT because TransportSimulateBulkActionIT doesn't have access to the data streams plugin, so it can't create data streams.

Reply:
> It's only possible to simulate index requests. I can do a null check just in case that changes at some point in the future.

What about adding an assertion? This way you indicate to another code maintainer that this is safe since only index requests can reach this point but it's also a light-weight safe guard in case something changes in the far far future.

Reply:
Perfect!

Reply:
I added a couple of assertions in there. There are also a couple asserting this in TransportSimulateBulkAction.